presto-main/src/main/java/com/facebook/presto/operator/MergeJoinOperator.java

```diff
@@ -0,0 +1,548 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.common.Page;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class MergeJoinOperator
+        implements Operator
+{
+    private final OperatorContext operatorContext;
+    private final List<Type> leftTypes;
+    private final List<Type> leftOutputTypes;
+    private final List<Integer> leftOutputChannels;
+    private final List<Type> rightTypes;
+    private final List<Type> rightOutputTypes;
+    private final List<Integer> rightOutputChannels;
+    private final List<Integer> leftJoinChannels;
+    private final List<Integer> rightJoinChannels;
+    private final MergeJoinSource mergeJoinSource;
+    private final JoinNode.Type joinType;
+
+    private MergeJoinPageBuilder outputPageBuilder;
+    private Page leftPage;
+    private Page rightPage;
+    private Optional<Match> leftMatch;
+    private Optional<Match> rightMatch;
+    private boolean finishing;
+    private boolean noMoreRightInput;
+    private int leftIndex;
+    private int rightIndex;
+
+    public MergeJoinOperator(
+            OperatorContext operatorContext,
+            List<Type> leftTypes,
+            List<Type> leftOutputTypes,
+            List<Integer> leftOutputChannels,
+            List<Type> rightTypes,
+            List<Type> rightOutputTypes,
+            List<Integer> rightOutputChannels,
+            List<Integer> leftJoinChannels,
+            List<Integer> rightJoinChannels,
+            JoinNode.Type joinType,
+            MergeJoinSource mergeJoinSource)
+    {
+        this.operatorContext = operatorContext;
+        this.leftTypes = leftTypes;
+        this.leftOutputTypes = leftOutputTypes;
+        this.leftOutputChannels = leftOutputChannels;
+        this.rightTypes = rightTypes;
+        this.rightOutputTypes = rightOutputTypes;
+        this.rightOutputChannels = rightOutputChannels;
+        this.leftJoinChannels = leftJoinChannels;
+        this.rightJoinChannels = rightJoinChannels;
+        this.joinType = joinType;
```

Comment:
although we already check in MergeJoinOperatorFactory, but for code integrity it's still better to do non null check here just in case other callers call this function
eg:
`this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");`